### PR TITLE
First version

### PR DESCRIPTION
**Pull request by Stradalec**

### CODE REVIEW

File: Lab6/Program.cs [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+          myMatrix.ShowMatrix(myMatrix);
+          continue;
+        }
+      }
+
+      Console.ReadKey();
+    }
+
+  }
+}
```
Comment (iredb) [inline Lab6/Program.cs]: используй стайлгайд от microsoft.

File: Lab6/Program.cs [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+      while (IsRunning) {
+        Console.WriteLine("Select your operation: 0 = summ, 1 = difference, 2 = compare, 3 = multiplication, " +
+                  "4 = reverse matrix, 5 = determinant, 6 = trace, 7 = transpOse, 8 = turn to diagonal");
+        select = Convert.ToInt32(Console.ReadLine());
+
+        eventSelector = new ChainApplication(myMatrix, select);
+        eventSelector.Run(1, myMatrix);
+
+        Console.WriteLine("Enough? y/n");
+        exit = Convert.ToChar(Console.ReadLine());
```
Comment (iredb) [inline Lab6/Program.cs]: подразумевается желание/нежелание выйти. почему переменная char?

**Review by iredb — CHANGES_REQUESTED** [review]

File: Lab6/Program.cs [hunk 3]
```diff
@@ -0,0 +1,710 @@
+﻿using System;
+
+namespace Lab6 {
+
+  //Пользовательский класс исключений
+  public class InvalidDepartmentException : System.Exception {
```
Comment (iredb) [inline Lab6/Program.cs]: use Microsoft style please. curly brackets must be on a new line.

File: Lab6/Program.cs (hunk 2 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Потому что вводится y "yes" или n "no". Следовательно, bool ставить не нужно, т.к. требуется проверять символ

File: Lab6/Program.cs (hunk 1 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Ладно. А что править?

File: Lab6/Program.cs (hunk 3 above)
Comment (iredb) [reply Lab6/Program.cs]: 1x tab = 4x spaces

File: Lab6/Program.cs [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+    : base(message) { }
+    public InvalidDepartmentException
+    (string message, System.Exception inner)
+    : base(message, inner) { }
+
+  }
+  public class Matrix : IComparable {
+    public double[,] InnerPartOfMatrix;
+    public int RowAndLineCount;
+    private double maxElement = 0;
```
Comment (iredb) [inline Lab6/Program.cs]: image

File: Lab6/Program.cs [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+    public InvalidDepartmentException(string message)
+    : base(message) { }
+    public InvalidDepartmentException
+    (string message, System.Exception inner)
+    : base(message, inner) { }
+
+  }
+  public class Matrix : IComparable {
+    public double[,] InnerPartOfMatrix;
+    public int RowAndLineCount;
```
Comment (iredb) [inline Lab6/Program.cs]: i think "rows and lines".
row == line == column???

File: Lab6/Program.cs [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+    : base() { }
+    public InvalidDepartmentException(string message)
+    : base(message) { }
+    public InvalidDepartmentException
+    (string message, System.Exception inner)
+    : base(message, inner) { }
+
+  }
+  public class Matrix : IComparable {
+    public double[,] InnerPartOfMatrix;
```
Comment (iredb) [inline Lab6/Program.cs]: array. should be "parts".

File: Lab6/Program.cs [hunk 7]
```diff
@@ -0,0 +1,710 @@
...
+
+      for (int rowIndex = 0; rowIndex < clonedRowCount.RowAndLineCount; ++rowIndex) {
+        for (int columnIndex = 0; columnIndex < clonedRowCount.RowAndLineCount; ++columnIndex) {
+          clonedRowCount.InnerPartOfMatrix[rowIndex, columnIndex] = this.InnerPartOfMatrix[rowIndex, columnIndex];
+        }
+      }
+
+      this.InnerPartOfMatrix[0, 0] = 0;
+
+      return clonedRowCount;
```
Comment (iredb) [inline Lab6/Program.cs]: why do matrix named "row count"?

File: Lab6/Program.cs [hunk 8]
```diff
@@ -0,0 +1,710 @@
...
+    }
+
+    // Приведение матрицы к диагональному виду
+    public void MatrixToDiagonal(Matrix inputMatrix) {
+
+      //Ищем главный элемент
+      for (int rowIndex = 0; rowIndex < inputMatrix.RowAndLineCount; ++rowIndex) {
+        maxElement = Math.Abs(inputMatrix.InnerPartOfMatrix[rowIndex, rowIndex]);
+        rowOfMaximumElement = rowIndex;
```
Comment (iredb) [inline Lab6/Program.cs]: row != rowIndex.
maybe _rowIndexOfMaximumElement or _maximumElementRowIndex?

File: Lab6/Program.cs [hunk 9]
```diff
@@ -0,0 +1,710 @@
...
+
+        for (int searchIndex = rowIndex + 1; searchIndex < inputMatrix.RowAndLineCount; ++searchIndex) {
+          if (Math.Abs(inputMatrix.InnerPartOfMatrix[searchIndex, rowIndex]) > maxElement) {
+            maxElement = Math.Abs(inputMatrix.InnerPartOfMatrix[searchIndex, rowIndex]);
```
Comment (iredb) [inline Lab6/Program.cs]: image

File: Lab6/Program.cs [hunk 10]
```diff
@@ -0,0 +1,710 @@
...
+
+    // Приведение матрицы к диагональному виду
+    public void MatrixToDiagonal(Matrix inputMatrix) {
+
+      //Ищем главный элемент
+      for (int rowIndex = 0; rowIndex < inputMatrix.RowAndLineCount; ++rowIndex) {
+        maxElement = Math.Abs(inputMatrix.InnerPartOfMatrix[rowIndex, rowIndex]);
+        rowOfMaximumElement = rowIndex;
+
+        for (int searchIndex = rowIndex + 1; searchIndex < inputMatrix.RowAndLineCount; ++searchIndex) {
```
Comment (iredb) [inline Lab6/Program.cs]: which index do you search?

File: Lab6/Program.cs [hunk 11]
```diff
@@ -0,0 +1,710 @@
...
+          if (Math.Abs(inputMatrix.InnerPartOfMatrix[searchIndex, rowIndex]) > maxElement) {
+            maxElement = Math.Abs(inputMatrix.InnerPartOfMatrix[searchIndex, rowIndex]);
+            rowOfMaximumElement = searchIndex;
+          }
+        }
+        //Перемещаем главный элемент на диагональ
+        if (rowOfMaximumElement != rowIndex) {
+          for (int columnIndex = rowIndex; columnIndex < inputMatrix.RowAndLineCount; ++columnIndex) {
+            double secondTemporary = inputMatrix.InnerPartOfMatrix[rowIndex, columnIndex];
```
Comment (iredb) [inline Lab6/Program.cs]: what instances does it contains? take it normal name.

File: Lab6/Program.cs [hunk 12]
```diff
@@ -0,0 +1,710 @@
...
+            inputMatrix.InnerPartOfMatrix[rowIndex, columnIndex] = inputMatrix.InnerPartOfMatrix[rowOfMaximumElement, columnIndex];
+            inputMatrix.InnerPartOfMatrix[rowOfMaximumElement, columnIndex] = secondTemporary;
+          }
+        }
+        // Непосредственно приведение матрицы к треугольному виду (ещё половина пути)
+        for (int firstDiagIndex = rowIndex + 1; firstDiagIndex < inputMatrix.RowAndLineCount; ++firstDiagIndex) {
```
Comment (iredb) [inline Lab6/Program.cs]: image

File: Lab6/Program.cs [hunk 13]
```diff
@@ -0,0 +1,710 @@
...
+          double result = -inputMatrix.InnerPartOfMatrix[firstDiagIndex, rowIndex] / inputMatrix.InnerPartOfMatrix[rowIndex, rowIndex];
+          for (int secondDiagIndex = rowIndex; secondDiagIndex < inputMatrix.RowAndLineCount; ++secondDiagIndex) {
+            if (rowIndex == secondDiagIndex) {
+              inputMatrix.InnerPartOfMatrix[firstDiagIndex, secondDiagIndex] = 0;
+            } else {
```
Comment (iredb) [inline Lab6/Program.cs]: "else" must be on a new line.

File: Lab6/Program.cs [hunk 14]
```diff
@@ -0,0 +1,710 @@
...
+              inputMatrix.InnerPartOfMatrix[firstDiagIndex, secondDiagIndex] += result * inputMatrix.InnerPartOfMatrix[rowIndex, secondDiagIndex];
+            }
+          }
+        }
+      }
+      // А теперь - обратный ход
+      for (int anotherSearchIndex = inputMatrix.RowAndLineCount - 1; anotherSearchIndex > 0; --anotherSearchIndex) {
```
Comment (iredb) [inline Lab6/Program.cs]: chto ti ishesh?????

File: Lab6/Program.cs [hunk 15]
```diff
@@ -0,0 +1,710 @@
...
+      resultMatrix.ShowMatrix(resultMatrix);
+      return resultMatrix;
+    }
+
+    //Оператор умножения
+    public static Matrix operator *(Matrix first, Matrix second) {
+      Matrix resultMatrix = (Matrix)first.Clone(first.RowAndLineCount);
+
+      for (int rowIndex = 0; rowIndex < first.RowAndLineCount; ++rowIndex) {
+        Console.Write("\n");
```
Comment (iredb) [inline Lab6/Program.cs]: for what?

File: Lab6/Program.cs [hunk 16]
```diff
@@ -0,0 +1,710 @@
...
+            ++check;
+          };
+        }
+      }
+      if (check == 0) {
+        return true;
+      } else {
+        return false;
+      }
+    }
```
Comment (iredb) [inline Lab6/Program.cs]: return check == 0

File: Lab6/Program.cs [hunk 17]
```diff
@@ -0,0 +1,710 @@
...
+    public static bool operator false(Matrix inputMatrix) {
+      return inputMatrix.RowAndLineCount == 1;
+    }
+
+    //Оператор CompareTo
+    int IComparable.CompareTo(object myObject) {
+      if (myObject is Matrix) {
+        var parameter = myObject as Matrix;
+        if (parameter.RowAndLineCount > this.RowAndLineCount) {
+          Console.WriteLine("1 case");
```
Comment (iredb) [inline Lab6/Program.cs]: is it for tests? why does this not "print" method do "console.wrteline"?

File: Lab6/Program.cs [hunk 18]
```diff
@@ -0,0 +1,710 @@
...
+
+    //Транспонирование
+    public static Matrix MatrixTranspose(this Matrix inputMatrix) {
+      for (int rowIndex = 0; rowIndex < inputMatrix.RowAndLineCount; ++rowIndex) {
+        Console.Write("\n");
+        for (int columnIndex = 0; columnIndex < inputMatrix.RowAndLineCount; ++columnIndex) {
+          Console.Write(inputMatrix.InnerPartOfMatrix[columnIndex, rowIndex] + " ");
+        }
+      }
+      return resultMatrix;
```
Comment (iredb) [inline Lab6/Program.cs]: where does it count?

File: Lab6/Program.cs [hunk 19]
```diff
@@ -0,0 +1,710 @@
...
+  class TransposeEvent : AncestorOfEvents {
+    public TransposeEvent() {
+      EventType = "Transpose";
+    }
+    public override void Call(Matrix inputMatrix) {
+      inputMatrix = inputMatrix.MatrixTranspose();
+    }
+  }
+
+  //Событие диагональной матрицы
```
Comment (iredb) [inline Lab6/Program.cs]: using descriptive names allows you to avoid writing comments

File: Lab6/Program.cs [hunk 20]
```diff
@@ -0,0 +1,710 @@
...
+  class LonelyEvent : AncestorOfEvents {
+    public LonelyEvent() {
+      EventType = "Lonely, no operation";
+    }
+  }
+
+  // Предок всех обработчиков
+  public abstract class AncestorOfHandler {
+    protected AncestorOfEvents innerEvent { get; set; }
+    protected AncestorOfHandler next { get; set; }
```
Comment (iredb) [inline Lab6/Program.cs]: image

image

File: Lab6/Program.cs [hunk 21]
```diff
@@ -0,0 +1,710 @@
...
+      innerEvent = new TransposeEvent();
+      next = new Handler9();
+    }
+  }
+  class Handler9 : AncestorOfHandler {
+    public Handler9() {
+      innerEvent = new DiagonalEvent();
+      next = null;
+    }
+  }
```
Comment (iredb) [inline Lab6/Program.cs]: image

File: Lab6/Program.cs [hunk 22]
```diff
@@ -0,0 +1,710 @@
...
+
+  //Класс "Цепь подчинения" (честно украдено из лекции)
+  public class ChainApplication {
+    private AncestorOfHandler eventHandler;
+    private int typeOfOperation;
```
Comment (iredb) [inline Lab6/Program.cs]: image

File: Lab6/Program.cs [hunk 23]
```diff
@@ -0,0 +1,710 @@
...
+    public ChainApplication(Matrix inputMatrix, int inputType) {
+      eventHandler = new Handler1(); typeOfOperation = inputType;
```
Comment (iredb) [inline Lab6/Program.cs]: write it in 2 lines, please

File: Lab6/Program.cs [hunk 24]
```diff
@@ -0,0 +1,710 @@
...
+      int select = 0;
+      ChainApplication eventSelector = new ChainApplication(myMatrix, select);
+      Matrix myCloneMatrix = (Matrix)myMatrix.Clone(myMatrix.RowAndLineCount);
+      Matrix resultMatrix = (Matrix)myMatrix.Clone(myMatrix.RowAndLineCount);
+      bool IsRunning = true;
+      char exit = ' ';
+
+      //Показываем пользователю матрицы
+      myMatrix.ShowMatrix(myMatrix);
+      myCloneMatrix.ShowMatrix(myCloneMatrix);
```
Comment (iredb) [inline Lab6/Program.cs]: why do myMatrix's method gets itself as argument?

File: Lab6/Program.cs [hunk 25]
```diff
@@ -0,0 +1,710 @@
...
+
+  //Делегат анонимной функции
+  public delegate void MatrixToDiag(Matrix inputMatrix);
+
+  internal class Program {
+
+    //Основная часть
+    static void Main(string[] args) {
+      Matrix myMatrix = new Matrix();
+      int select = 0;
```
Comment (iredb) [inline Lab6/Program.cs]: select is verb. use nouns for variables.

File: Lab6/Program.cs [hunk 26]
```diff
@@ -0,0 +1,710 @@
...
+      ChainApplication eventSelector = new ChainApplication(myMatrix, select);
+      Matrix myCloneMatrix = (Matrix)myMatrix.Clone(myMatrix.RowAndLineCount);
+      Matrix resultMatrix = (Matrix)myMatrix.Clone(myMatrix.RowAndLineCount);
+      bool IsRunning = true;
```
Comment (iredb) [inline Lab6/Program.cs]: image

File: Lab6/Program.cs (hunk 2 above)
Comment (iredb) [reply Lab6/Program.cs]: сделай bool.

File: Lab6/Program.cs [hunk 27]
```diff
@@ -0,0 +1,710 @@
...
+
+        eventSelector = new ChainApplication(myMatrix, select);
+        eventSelector.Run(1, myMatrix);
+
+        Console.WriteLine("Enough? y/n");
+        exit = Convert.ToChar(Console.ReadLine());
+        if (exit == 'y') {
+          Console.WriteLine("Okay. Press any button to exit completely");
+          IsRunning = false;
+        } else {
```
Comment (iredb) [inline Lab6/Program.cs]: 'Y' == 'n'

Comment (iredb) [reply Lab6/Program.cs]: "y" with caps lock is other symbol. i think, if i enter "Y", expression in "if()" whould be false.

**Review by iredb — CHANGES_REQUESTED** [review]

File: Lab6/Program.cs (hunk 2 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Okay....

File: Lab6/Program.cs (hunk 3 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Accepted, I am trying to fix that

File: Lab6/Program.cs (hunk 4 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Accepted

File: Lab6/Program.cs (hunk 5 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Accepted. My matrix always square, so row = column

File: Lab6/Program.cs (hunk 6 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Accepted

File: Lab6/Program.cs (hunk 7 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Idk, I will fix that

File: Lab6/Program.cs (hunk 8 above)
Comment (Stradalec) [reply Lab6/Program.cs]: You are right, I must rename this. Accepted

File: Lab6/Program.cs (hunk 9 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Definitely true. Accepted

File: Lab6/Program.cs (hunk 10 above)
Comment (Stradalec) [reply Lab6/Program.cs]: SearchMaximumElementIndex. Accepted

File: Lab6/Program.cs (hunk 11 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Accepted

File: Lab6/Program.cs (hunk 12 above)
Comment (Stradalec) [reply Lab6/Program.cs]: True. Accepted

File: Lab6/Program.cs (hunk 13 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Okay

File: Lab6/Program.cs (hunk 14 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Happiness.....
Fixed

File: Lab6/Program.cs (hunk 15 above)
Comment (Stradalec) [reply Lab6/Program.cs]: For some reasons this does not work without that

File: Lab6/Program.cs (hunk 16 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Okay

File: Lab6/Program.cs (hunk 17 above)
Comment (Stradalec) [reply Lab6/Program.cs]: It is from third laba. This is a monument.

File: Lab6/Program.cs (hunk 18 above)
Comment (Stradalec) [reply Lab6/Program.cs]: It just transpose, no need to count, just reverse indexes

File: Lab6/Program.cs (hunk 19 above)
Comment (Stradalec) [reply Lab6/Program.cs]: You are rught, but I want to write 100500 comments in my code

File: Lab6/Program.cs (hunk 20 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Okay

File: Lab6/Program.cs (hunk 21 above)
Comment (Stradalec) [reply Lab6/Program.cs]: This naming help me to undestand queue of operations. But teamleader == god of this place, so I am going to rename handlers

File: Lab6/Program.cs (hunk 22 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Ok

File: Lab6/Program.cs (hunk 23 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Okay, I will

File: Lab6/Program.cs (hunk 24 above)
Comment (Stradalec) [reply Lab6/Program.cs]: I dont know, really

Comment (Stradalec) [reply Lab6/Program.cs]: I am tried to fix that, but without this ShowMatrix() method does not work in Main

File: Lab6/Program.cs (hunk 25 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Okay

File: Lab6/Program.cs (hunk 26 above)
Comment (Stradalec) [reply Lab6/Program.cs]: True

File: Lab6/Program.cs (hunk 27 above)
Comment (Stradalec) [reply Lab6/Program.cs]: Okay, fixed

**Pull request closed without merging**